### ISSUE TITLE
More forgiving parsing of AVD files requested

### ISSUE THREAD

**Issue by alesimula** [zathras/jovial_svg#13]
Hello, correct me if I'm wrong, but this library does not seem to have any kind of fault tolerance

I was trying to read an Android Vector Drawable with ScalableImage.fromAvdString, but even with warn set to false, if an unrecognized attribute is found, it will fail with Exception: PathError(Unexpected attribute WHATEVER)

Is there no way to add fault tolerance? (skip unrecognized attributes or values instead of failing); if not, then I'm suggesting for it to be implemented (something like ScalableImage.fromAvdString(xmlString, faultTolerant: true))

**Comment by alesimula** [comment]
@zathras Just to make an example here is an xml that would fail for multiple reason

```
<?xml version="1.0" encoding="utf-8"?>
<vector xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:height="108dip" android:width="108dip" android:viewportWidth="108" android:viewportHeight="108">
    <group android:scaleX="0.228975" android:scaleY="0.228975" android:translateX="21.6" android:translateY="21.6">
        <group>
            <clip-path android:pathData="M0,0h283v283h-283z"/>
            <path android:fillColor="#ffeeeeee" android:pathData="M134.975,56.317C135.235,56.087 135.511,55.869 135.802,55.665C139.966,52.749 145.705,53.761 148.621,57.925L191.087,118.573H190.655H168.614L140.84,78.907L112.574,119.408H90.334L132.857,58.478C133.455,57.62 134.174,56.898 134.975,56.317Z" android:fillType="evenOdd"/>
            <path android:fillColor="#fff487ac" android:pathData="M146.532,150C148.679,144.584 149.859,138.68 149.859,132.5C149.859,119.928 144.975,108.497 137,100H198.859C207.144,100 213.859,106.716 213.859,115V150H146.532Z" android:fillType="evenOdd"/>
            <path  android:pathData="M129,107.778l7.778,-7.778l22.05,22.05l-7.778,7.778z">
                <aapt:attr name="android:fillColor"><gradient xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:angle="0" android:type="linear" android:startX="42.9492" android:startY="49.5979" android:endX="85.8476" android:endY="92.4963">
                    <item android:color="#44000000" android:offset="0"/>
                    <item android:color="#00000000" android:offset="1"/>
                </gradient>
                </aapt:attr>
            </path>
            <path android:fillColor="#ff80cbc4" android:pathData="M102.5,132.5m-47.5,0a47.5,47.5 0,1 1,95 0a47.5,47.5 0,1 1,-95 0"/>
            <path android:fillColor="#fffafafa" android:pathData="M146.389,55.704C142.274,52.823 136.602,53.823 133.72,57.938L91.263,118.573H113.472L148.623,68.373C151.504,64.258 150.504,58.586 146.389,55.704Z" android:fillType="evenOdd"/>
            <path android:fillColor="#fff5f5f5" android:pathData="M146.337,55.839C142.024,52.81 136.083,53.857 133.066,58.179L90.335,119.408H113.611L148.684,69.152C151.701,64.829 150.65,58.869 146.337,55.839Z" android:fillType="evenOdd"/>
            <path android:fillColor="#ffe4e4e4" android:pathData="M115.989,116H92.714L92,117.023V119H113.896L115.989,116Z" android:fillType="evenOdd"/>
            <path android:fillColor="#fffafafa" android:pathData="M187.837,117H231.818C236.868,117 241,121.125 241,126.167L240.725,128.642L217.403,213.617C215.199,221.317 208.129,227 199.682,227H80.318C71.871,227 64.801,221.317 62.689,213.617L39.367,128.642C39.092,127.817 39,126.992 39,126.167C39,121.125 43.132,117 48.182,117H187.837Z"/>
            <path  android:pathData="M132.726,144.539L105.577,191.185C106.389,190.008 107.478,189.046 108.751,188.382C110.028,187.715 111.451,187.369 112.894,187.374L128.183,187.374H128.184L140.982,165.383L133.338,152.244C132.648,151.091 132.235,149.797 132.129,148.46C132.023,147.125 132.227,145.783 132.726,144.539ZM148.834,151.884C148.87,151.816 148.906,151.748 148.941,151.68C148.975,151.61 149.009,151.541 149.042,151.47C149.167,151.206 149.278,150.936 149.375,150.66C149.473,150.385 149.557,150.104 149.626,149.82C149.695,149.536 149.749,149.248 149.789,148.958C149.818,148.755 149.84,148.552 149.854,148.347C149.859,148.264 149.864,148.181 149.867,148.098C149.855,148.386 149.829,148.673 149.789,148.959C149.749,149.248 149.694,149.536 149.626,149.82C149.556,150.104 149.472,150.385 149.374,150.66C149.277,150.936 149.166,151.206 149.042,151.47C148.976,151.61 148.907,151.748 148.834,151.884ZM149.868,148.072C149.871,148.004 149.872,147.936 149.873,147.868C149.873,147.855 149.873,147.843 149.873,147.83C149.873,147.911 149.871,147.992 149.868,148.072Z" android:fillType="evenOdd">
                <aapt:attr name="android:fillColor"><gradient xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:angle="0" android:type="linear" android:startX="137.925" android:startY="147.366" android:endX="126.884" android:endY="185.455">
                    <item android:color="#ffff80ab" android:offset="0"/>
                    <item android:color="#ffff4081" android:offset="1"/>
                </gradient>
                </aapt:attr>
            </path>
            <path android:fillColor="#ff00b0ff" android:pathData="M105.567,191.2C104.969,192.069 104.533,193.038 104.281,194.061C104.343,193.816 104.416,193.574 104.499,193.336C104.597,193.06 104.709,192.79 104.835,192.526C104.959,192.261 105.096,192.003 105.246,191.752L105.567,191.2Z"/>
            <path  android:pathData="M114.122,204.894H167.792C167.765,204.891 167.738,204.888 167.71,204.885C167.42,204.838 167.132,204.777 166.848,204.703C166.563,204.627 166.283,204.536 166.007,204.433C165.732,204.33 165.463,204.213 165.199,204.084C164.935,203.953 164.678,203.809 164.428,203.653C164.18,203.498 163.939,203.331 163.708,203.152C163.476,202.972 163.252,202.781 163.039,202.579C162.826,202.378 162.623,202.166 162.431,201.945C162.239,201.724 162.059,201.493 161.89,201.253C161.721,201.013 161.565,200.765 161.421,200.51L153.779,187.374H128.183L120.539,200.51C120.437,200.69 120.329,200.866 120.215,201.038C120.054,201.282 119.882,201.518 119.698,201.745C119.512,201.972 119.315,202.189 119.108,202.397C118.901,202.605 118.684,202.802 118.457,202.989C118.23,203.175 117.994,203.35 117.749,203.513C117.505,203.676 117.253,203.827 116.993,203.965C116.733,204.103 116.467,204.227 116.194,204.339C115.922,204.449 115.644,204.547 115.362,204.63C115.081,204.714 114.795,204.784 114.506,204.839C114.379,204.86 114.251,204.878 114.122,204.894Z">
                <aapt:attr name="android:fillColor"><gradient xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:angle="0" android:type="linear" android:startX="135.91" android:startY="187.374" android:endX="135.91" android:endY="204.894">
                    <item android:color="#ff81d4fa" android:offset="0"/>
                    <item android:color="#ff4fc3f7" android:offset="1"/>
                </gradient>
                </aapt:attr>
            </path>
            <path  android:pathData="M105.567,191.2L105.246,191.752C105.096,192.003 104.959,192.261 104.835,192.526C104.709,192.79 104.597,193.06 104.499,193.336C104.416,193.574 104.343,193.816 104.281,194.061C104.256,194.164 104.233,194.269 104.211,194.373C104.185,194.498 104.16,194.623 104.139,194.749C103.936,196.004 104.011,197.287 104.359,198.51C104.706,199.733 105.318,200.867 106.15,201.832C106.983,202.797 108.018,203.571 109.182,204.099C110.347,204.628 111.613,204.899 112.894,204.894L114.122,204.894C114.251,204.878 114.379,204.86 114.506,204.839C114.795,204.784 115.081,204.714 115.362,204.63C115.644,204.547 115.922,204.449 116.194,204.339C116.467,204.227 116.733,204.103 116.993,203.965C117.253,203.827 117.505,203.676 117.749,203.513C117.994,203.35 118.23,203.175 118.457,202.989C118.684,202.802 118.901,202.605 119.108,202.397C119.315,202.189 119.512,201.972 119.698,201.745C119.882,201.518 120.054,201.282 120.215,201.038C120.329,200.866 120.437,200.69 120.539,200.51L128.183,187.374L112.894,187.374C111.45,187.369 110.028,187.715 108.75,188.382C107.473,189.049 106.38,190.017 105.567,191.2Z">
                <aapt:attr name="android:fillColor"><gradient xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:angle="0" android:type="linear" android:startX="135.91" android:startY="187.374" android:endX="135.91" android:endY="204.894">
                    <item android:color="#ff81d4fa" android:offset="0"/>
                    <item android:color="#ff4fc3f7" android:offset="1"/>
                </gradient>
                </aapt:attr>
            </path>
            <path android:fillColor="#ff40c4ff" android:pathData="M153.779,187.374L161.421,200.51C161.565,200.766 161.721,201.014 161.89,201.254C162.059,201.493 162.239,201.724 162.431,201.945C162.623,202.167 162.826,202.378 163.039,202.579C163.252,202.781 163.476,202.973 163.708,203.153C163.939,203.331 164.18,203.498 164.428,203.653C164.678,203.81 164.935,203.953 165.199,204.084C165.463,204.214 165.732,204.33 166.007,204.433C166.283,204.537 166.563,204.627 166.848,204.703C167.132,204.778 167.42,204.839 167.71,204.885C167.737,204.888 167.765,204.891 167.792,204.894H169.069C170.672,204.913 172.251,204.499 173.635,203.695C175.019,202.891 176.157,201.728 176.925,200.332C176.971,200.249 177.015,200.166 177.058,200.083C177.113,199.972 177.165,199.859 177.215,199.746C177.311,199.534 177.397,199.319 177.476,199.1C177.572,198.823 177.655,198.542 177.724,198.257C177.791,197.973 177.845,197.687 177.884,197.398C177.923,197.108 177.947,196.817 177.957,196.524C177.965,196.232 177.959,195.939 177.938,195.648C177.919,195.357 177.884,195.067 177.836,194.78C177.786,194.492 177.723,194.206 177.645,193.924C177.567,193.643 177.475,193.365 177.37,193.093C177.264,192.82 177.145,192.553 177.013,192.292C176.919,192.109 176.819,191.929 176.713,191.752L176.123,190.736C175.287,189.675 174.215,188.82 172.992,188.237C171.768,187.654 170.426,187.359 169.069,187.375L153.779,187.374Z"/>
            <path android:fillColor="#ffffa943" android:pathData="M132.727,144.535L133.338,143.486C133.437,143.31 133.542,143.138 133.653,142.97C133.814,142.725 133.988,142.488 134.173,142.26C134.358,142.033 134.554,141.816 134.76,141.608C134.967,141.4 135.184,141.203 135.41,141.016C135.637,140.83 135.874,140.655 136.118,140.492C136.363,140.328 136.615,140.178 136.875,140.039C137.133,139.902 137.399,139.778 137.67,139.666C137.942,139.555 138.22,139.456 138.502,139.372C138.785,139.288 139.072,139.218 139.361,139.163C139.647,139.107 139.936,139.066 140.226,139.038C138.584,139.169 137.011,139.753 135.686,140.725C134.36,141.696 133.336,143.016 132.727,144.535Z"/>
            <path android:fillColor="#ffffa943" android:pathData="M140.889,139.011C140.784,139.008 140.679,139.007 140.574,139.008C140.463,139.016 140.352,139.026 140.242,139.038C140.457,139.021 140.673,139.012 140.889,139.011V139.011Z"/>
            <path  android:pathData="M140.241,139.038C140.457,139.021 140.673,139.012 140.889,139.011L140.888,139.011C141.05,139.008 141.213,139.01 141.375,139.017C141.426,139.019 141.476,139.021 141.526,139.023C142.991,139.122 144.408,139.582 145.648,140.362C146.888,141.141 147.913,142.215 148.63,143.486L149.265,144.575C149.276,144.601 149.287,144.627 149.298,144.654C149.403,144.926 149.494,145.204 149.57,145.485C149.646,145.768 149.708,146.054 149.755,146.343C149.803,146.631 149.836,146.92 149.855,147.211C149.874,147.503 149.878,147.796 149.867,148.088C149.856,148.379 149.829,148.67 149.788,148.959C149.749,149.248 149.694,149.536 149.625,149.82C149.556,150.104 149.472,150.385 149.374,150.66C149.277,150.936 149.166,151.206 149.041,151.47C148.917,151.735 148.78,151.993 148.63,152.244L140.982,165.383L133.337,152.244C132.648,151.091 132.234,149.797 132.128,148.46C132.022,147.124 132.227,145.781 132.726,144.535L133.337,143.486C133.436,143.31 133.541,143.138 133.652,142.97C133.813,142.725 133.987,142.488 134.172,142.26C134.357,142.033 134.553,141.815 134.759,141.608C134.966,141.4 135.183,141.203 135.41,141.016C135.637,140.83 135.873,140.655 136.118,140.491C136.362,140.328 136.614,140.178 136.874,140.039C137.133,139.902 137.398,139.777 137.67,139.666C137.942,139.554 138.219,139.456 138.501,139.372C138.784,139.288 139.071,139.218 139.361,139.163C139.647,139.107 139.936,139.065 140.226,139.038H140.241ZM149.299,144.654C149.288,144.627 149.277,144.601 149.265,144.575L176.713,191.752C177.305,192.749 177.692,193.852 177.852,194.998C178.012,196.144 177.941,197.31 177.645,198.429C177.348,199.548 176.831,200.598 176.123,201.517C175.415,202.436 174.53,203.208 173.52,203.786C172.511,204.364 171.395,204.739 170.239,204.887C169.082,205.035 167.908,204.955 166.783,204.65C165.658,204.346 164.605,203.823 163.684,203.113C162.764,202.402 161.995,201.518 161.421,200.51L140.983,165.383L148.63,152.244C148.78,151.993 148.918,151.735 149.042,151.47C149.166,151.206 149.277,150.936 149.375,150.66C149.472,150.385 149.556,150.104 149.626,149.82C149.695,149.536 149.749,149.248 149.789,148.958C149.83,148.67 149.856,148.379 149.868,148.088C149.878,147.796 149.874,147.503 149.855,147.211C149.837,146.92 149.803,146.631 149.756,146.343C149.709,146.054 149.647,145.768 149.571,145.485C149.494,145.203 149.403,144.926 149.299,144.654Z" android:fillType="evenOdd">
                <aapt:attr name="android:fillColor"><gradient xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:angle="0" android:type="linear" android:startX="138" android:startY="143.5" android:endX="173.5" android:endY="202">
                    <item android:color="#ffffc400" android:offset="0"/>
                    <item android:color="#ffffb300" android:offset="1"/>
                </gradient>
                </aapt:attr>
            </path>
        </group>
    </group>
</vector>
```

This xml was generated by decompiling an apk, then combining the xml of the app icon with the xml of the gradients of four of its paths; for that reason, in this case those gradients contain the xmlns:android and xmlns:aapt attributes (which I had to remove with a regular expression to make it wark)

Additionally, the android:angle attribute does not seem to be recognized and I had to remove that too.

I know this is a very specific scenario, but in general, I think any unrecognised attribute should just be ignored

On top of that, if a property contains an unrecognized value (eg. android:type="nope") the library behaves by failing completely rather than 

a) applying the default value to that attribute
b) ignoring that attribute
c) not rendering only the path in question instead of the whole icon

I also think Android accepts numerical values on the enum attributes as valid (at least, Android Studio seems to accept them and the decompiled xml originally had numerical values on all its enum attributes), eg. android:type="0", where 0 is "linear" or android:fillType="1" where 1 is "evenOdd", this was also something I had to replace with regular expression in order to be able to read the icon correctly, here is how I mapped these values to replace them (there are probably more)

```
Map<String, String> fillType = {
  "0": "winding",
  "1": "evenOdd",
  "2": "inverseWinding",
  "3": "inverseEvenOdd",
};
Map<String, String> gradientType = {
  "0": "linear",
  "1": "radial",
  "2": "sweep"
};
```

**Comment by zathras** [comment]
Oh, you're talking about AVD files.  You really buried the lede there!  Actually, more than buried, but OK, now I know you're not talking about SVG.  I updated the RFE description to something more informative.

So, what you're asking for is an option for *AVD* file parsing where unrecognized attributes and unrecognized property values are ignored.  Is that a complete description of what you're suggesting?  Do you have more sample AVD files?  (Attachments are preferred).  The more complete a set of examples are, the better the chances that the code will hit all the relevant cases.

**Comment by alesimula** [comment]
@zathras Yep, that's what I'm proposing, I think having a flag to tell the parser whether to be fault tolerant would be the best course of action.

Here is an xml with some garbage data (couldn't put it as an attachment, sorry), along with a screenshot that shows how it's (correctly) rendered on Android Studio

```
<?xml version="1.0" encoding="utf-8"?>
<vector xmlns:android="http://schemas.android.com/apk/res/android" xmlns:aapt="http://schemas.android.com/aapt" android:height="108dip" android:width="108dip" android:viewportWidth="108" android:viewportHeight="108">
    <GARBAGE>
        loose_garbage
    </GARBAGE>
    <group SOMEGARBAGE android:scaleX="0.228975" android:scaleY="0.228975" android:translateX="21.6" android:translateY="21.6">
        <group>
            <clip-path MOREGARBAGE android:pathData="M0,0h283v283h-283z"/>
            <GARBAGEELEMENT GARBAGEATTRIBUTE='GARBAGEVALUE'></GARBAGEELEMENT>
            <path android:fillType="GARBAGE_VALUE" android:fillColor="#FFEEEEEE" android:pathData="M134.975,56.317C135.235,56.087 135.511,55.869 135.802,55.665C139.966,52.749 145.705,53.761 148.621,57.925L191.087,118.573H190.655H168.614L140.84,78.907L112.574,119.408H90.334L132.857,58.478C133.455,57.62 134.174,56.898 134.975,56.317Z"/>
            <path android:fillColor="#FFF487AC" android:pathData="M146.532,150C148.679,144.584 149.859,138.68 149.859,132.5C149.859,119.928 144.975,108.497 137,100H198.859C207.144,100 213.859,106.716 213.859,115V150H146.532Z" android:fillType="evenOdd"/>
            <path android:fillColor="#FF80CBC4" android:pathData="M102.5,132.5m-47.5,0a47.5,47.5 0,1 1,95 0a47.5,47.5 0,1 1,-95 0"/>
            <path android:fillColor="#FFFAFAFA" android:pathData="M146.389,55.704C142.274,52.823 136.602,53.823 133.72,57.938L91.263,118.573H113.472L148.623,68.373C151.504,64.258 150.504,58.586 146.389,55.704Z" android:fillType="evenOdd"/>
            <path android:fillColor="#FFF5F5F5" android:pathData="M146.337,55.839C142.024,52.81 136.083,53.857 133.066,58.179L90.335,119.408H113.611L148.684,69.152C151.701,64.829 150.65,58.869 146.337,55.839Z" android:fillType="evenOdd"/>
            <path android:fillColor="#FFE4E4E4" android:pathData="M115.989,116H92.714L92,117.023V119H113.896L115.989,116Z" android:fillType="evenOdd"/>
            <path android:fillColor="#FFFAFAFA" android:pathData="M187.837,117H231.818C236.868,117 241,121.125 241,126.167L240.725,128.642L217.403,213.617C215.199,221.317 208.129,227 199.682,227H80.318C71.871,227 64.801,221.317 62.689,213.617L39.367,128.642C39.092,127.817 39,126.992 39,126.167C39,121.125 43.132,117 48.182,117H187.837Z"/>
            </group>
    </group>
</vector>
```

![image](https://user-images.githubusercontent.com/3057584/143072887-699f3160-d609-4c39-9cfa-2d8bc240191e.png)

**Comment by zathras** [comment]
OK, thanks.  Makes sense.  A priori, this doesn't look like it should be too hard.  Is this urgent?

**Comment by alesimula** [comment]
@zathras Nope, I already managed to work around this "limitation", I'm just proposing this as I think it would make sense and benefit the library

**Comment by zathras** [comment]
OK, cool.  Thanks.

**Comment by zathras** [comment]
Hmmm...  Some of the errors in the second example make the file invalid XML.  That causes a parse exception from the XML parsing library I'm using.

I'm not too surprised that Android (or maybe just some of the Android tools?) are super forgiving with mal-formed XML, given the. hacky and under-specified nature of AVD files.  But doing surgery on XML parsing to handle this seems overkill.

However, the spirit of the RFE is sound:  There's no good reason to be strict with AVD parsing, so I'll turn errors into warnings for things like unrecognized attributes and tags.

**Comment by zathras** [comment]
Released in 1.0.8.  Thanks!
